File: robotics-codes-from-scratch-master/python/LQT.py
'''
Linear quadratic tracking (LQT) with a viapoints example

Copyright (c) 2023 Idiap Research Institute <https://www.idiap.ch/>
Written by Sylvain Calinon <https://calinon.ch>

This file is part of RCFS <https://rcfs.ch/>
License: MIT
'''

import numpy as np
from math import factorial
import matplotlib.pyplot as plt

# Parameters
# ===============================
param = lambda: None # Lazy way to define an empty class in python
param.dt = 1E-2 # Time step length
param.nbPoints = 2 # Number of viapoints
param.nbDeriv = 2 # Order of the dynamical system
param.nbVarPos = 2 # Number of position variable
param.nbVar = param.nbVarPos * param.nbDeriv # Dimension of state vector
param.nbData = 100  # Number of datapoints
param.rfactor = 1e-9  # Control weight term

R = np.eye((param.nbData-1) * param.nbVarPos) * param.rfactor  # Control cost matrix

# Defin the viapoint passe time 
tl = np.linspace(0, param.nbData, param.nbPoints+1)
tl = np.rint(tl[1:]).astype(np.int64) - 1 

idx = np.array([i + np.arange(0,param.nbVar,1) for i in (tl*param.nbVar)]).flatten() 

# Viapoints
MuTmp = np.vstack((np.random.rand(param.nbVarPos, param.nbPoints) * 5, np.zeros(((param.nbVar-param.nbVarPos), param.nbPoints))))
param.Mu = MuTmp.reshape((param.nbPoints*param.nbVar,1), order='F')
print(MuTmp)
print(param.Mu.shape)
# Precision matrix
Qtmp = np.diag(np.hstack((np.ones(param.nbVarPos), np.zeros(param.nbVar-param.nbVarPos))));
Q = np.kron(np.eye(param.nbPoints), Qtmp)


# Dynamical System settings (discrete)
# =====================================
A1d = np.zeros((param.nbDeriv,param.nbDeriv))
B1d = np.zeros((param.nbDeriv,1))

for i in range(param.nbDeriv):
    A1d += np.diag(np.ones(param.nbDeriv-i), i) * param.dt**i 
    B1d[param.nbDeriv-i-1] = param.dt**(i+1) 


# A = np.eye(param.nbv)
A = np.kron(A1d,np.eye(param.nbVarPos))
B = np.kron(B1d,np.eye(param.nbVarPos))
print(A)
print(B)

# Build Sx and Su transfer matrices
Su0 = np.zeros((param.nbVar*param.nbData,param.nbVarPos * (param.nbData-1))) 
Sx0 = np.kron(np.ones((param.nbData,1)),np.eye(param.nbVar,param.nbVar))
M = B
for i in range(1,param.nbData):
    Sx0[i*param.nbVar:param.nbData*param.nbVar,:] = np.dot(Sx0[i*param.nbVar:param.nbData*param.nbVar,:], A)
    Su0[param.nbVar*i:param.nbVar*i+M.shape[0],0:M.shape[1]] = M
    M = np.hstack((np.dot(A,M),B))

Su = Su0[idx,:] # Keeping rows of interest
Sx = Sx0[idx,:] # Keeping rows of interest


# Batch LQR Reproduction
# =====================================
x0 = np.zeros((param.nbVar,1))
u_hat = np.linalg.inv(Su.T @ Q @ Su + R) @ Su.T @ Q @ (param.Mu - Sx @ x0)
x_hat = Sx0 @ x0 + Su0 @ u_hat


# Plotting
# =========
plt.figure()
plt.title("2D Trajectory")
plt.scatter(x_hat[0],x_hat[1],c='black',s=100)
plt.scatter(param.Mu[::param.nbVar],param.Mu[1::param.nbVar],c='red',s=100)  
plt.plot(x_hat[::param.nbVar] , x_hat[1::param.nbVar], c='black')
plt.axis("off")
plt.gca().set_aspect('equal', adjustable='box')

fig,axs = plt.subplots(2,1)
axs[0].scatter(tl,param.Mu[0::param.nbVar],c='red')  
axs[0].plot(x_hat[::param.nbVar], c='black')
axs[0].set_ylabel("$x_1$")
axs[0].set_xticks([0,param.nbData])
axs[0].set_xticklabels(["0","T"])

axs[1].scatter(tl,param.Mu[1::param.nbVar],c='red')  
axs[1].plot(x_hat[1::param.nbVar], c='black')
axs[1].set_ylabel("$x_2$")
axs[1].set_xlabel("$t$")
axs[1].set_xticks([0,param.nbData])
axs[1].set_xticklabels(["0","T"])

plt.show()

File: robotics-codes-from-scratch-master/matlab/LQT.m
%% Linear quadratic tracking (LQT) applied to a viapoint task (batch formulation)
%% 
%% Copyright (c) 2023 Idiap Research Institute <https://www.idiap.ch/>
%% Written by Sylvain Calinon <https://calinon.ch>
%% 
%% This file is part of RCFS <https://robotics-codes-from-scratch.github.io/>
%% License: MIT

function LQT

%% Parameters
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
param.nbData = 100; %Number of datapoints
param.nbPoints = 3; %Number of viapoints
param.nbVarPos = 2; %Dimension of position data (here: x1,x2)
param.nbDeriv = 2; %Number of static and dynamic features (nbDeriv=2 for [x,dx] and u=ddx)
param.nbVar = param.nbVarPos * param.nbDeriv; %Dimension of state vector
param.dt = 1E-2; %Time step duration
param.r = 1E-12; %Control cost in LQR (for other cases)

%Task setting (viapoints passing)
Mu = [rand(param.nbVarPos, param.nbPoints) - 0.5; zeros(param.nbVar-param.nbVarPos, param.nbPoints)]; %Viapoints

%Q = speye(param.nbVar * param.nbPoints) * 1E0; %Precision matrix (for full state)
Q = kron(eye(param.nbPoints), diag([ones(1,param.nbVarPos) * 1E0, zeros(1,param.nbVar-param.nbVarPos)])); %Precision matrix (for position only)

R = speye((param.nbData-1)*param.nbVarPos) * param.r; %Standard control weight matrix (at trajectory level)
%e = ones(param.nbData-1,1) * param.r;
%R = kron(spdiags([-e 2*e -e], -1:1, param.nbData-1, param.nbData-1), speye(param.nbVarPos)); %Control weight matrix as transition matrix to encourage smoothness

%Time occurrence of viapoints
tl = linspace(1, param.nbData, param.nbPoints+1);
tl = round(tl(2:end));
idx = (tl - 1) * param.nbVar + [1:param.nbVar]';

%Dynamical System settings (discrete version)
[Su0, Sx0] = computeSuSx(param);
Su = Su0(idx,:);
Sx = Sx0(idx,:);


%% LQT (batch formulation)
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
x0 = zeros(param.nbVar,1); %Initial state
u = (Su' * Q * Su + R) \ Su' * Q * (Mu(:) - Sx * x0); %Control commands
rx = reshape(Sx0*x0+Su0*u, param.nbVar, param.nbData); %Reproduced trajectory


%%% LQT with a new time step recomputed based on the maximum allowed velocity
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%dx_max = 5.0; %Allowed velocity limit
%[dx, id] = max(max(abs(rx(param.nbVarPos+1:2*param.nbVarPos,:)))); %Determine highest velocity
%s = dx / dx_max; %Compute scaling factor
%param.dt = param.dt * s; %New time step
%u = u / s^param.nbDeriv; %New control commands
%[Su0, Sx0] = computeSuSx(param); %New linear system
%rx = reshape(Sx0*x0+Su0*u, param.nbVar, param.nbData); %New reproduced trajectory


%% Plot 2D
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
h(1) = figure('position',[10 10 800 800]); hold on; axis off;
plot(rx(1,:), rx(2,:), 'k-','linewidth',2);
plot(rx(1,1), rx(2,1), 'k.','markersize',30);
plot(Mu(1,:), Mu(2,:), 'r.','markersize',30);
axis equal; 


%% Timeline plot
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
h(2) = figure('position',[950 10 800 800],'color',[1 1 1]); 
for j=1:param.nbVarPos
	subplot(param.nbVarPos,1,j); hold on;
	plot(rx(j,:), 'k-','linewidth',2);
	plot(rx(j,1), 'k.','markersize',30);
	plot(tl, Mu(j,:), 'r.','markersize',30);
	xlabel('t','fontsize',26); ylabel(['x_' num2str(j)],'fontsize',26);
end

waitfor(h);
end


function [Su0, Sx0] = computeSuSx(param)
    %Dynamical System settings (discrete version)
    A1d = zeros(param.nbDeriv);
    for i=0:param.nbDeriv-1
	    A1d = A1d + diag(ones(param.nbDeriv-i,1),i) * param.dt^i * 1/factorial(i); %Discrete 1D
    end
    B1d = zeros(param.nbDeriv,1); 
    for i=1:param.nbDeriv
	    B1d(param.nbDeriv-i+1) = param.dt^i * 1/factorial(i); %Discrete 1D
    end
    A = kron(A1d, speye(param.nbVarPos)); %Discrete nD
    B = kron(B1d, speye(param.nbVarPos)); %Discrete nD

    %Build Sx and Su transfer matrices
    Sx0 = kron(ones(param.nbData,1), speye(param.nbVar));
    Su0 = sparse(param.nbVar*param.nbData, param.nbVarPos*(param.nbData-1));
    M = B;
    for n=2:param.nbData
	    id1 = (n-1)*param.nbVar+1:param.nbData*param.nbVar;
	    Sx0(id1,:) = Sx0(id1,:) * A;
	    id1 = (n-1)*param.nbVar+1:n*param.nbVar; 
	    id2 = 1:(n-1)*param.nbVarPos;
	    Su0(id1,id2) = M;
	    M = [A*M(:,1:param.nbVarPos), M]; 
    end
end
